### PR TITLE
Yt/moist roe

### PR DESCRIPTION
**Pull request by slifer50**
# Description

An Moist Roe flux implementation

I have added a Moist Roe flux, a moist rising bubble test and Riemann shock tube type test. Also I have added additional boundary conditions necessary for running these test.

- [ ] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [ ] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [ ] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [ ] There are no open pull requests for this already
- [ ] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -23,6 +23,7 @@ Formatting = "59287772-0a20-5a39-b81b-1366585eb4c0"
 ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210"
 GaussQuadrature = "d54b0c1a-921d-58e0-8e36-89d8069c0969"
 JLD2 = "033835bb-8acc-5ee8-8aae-3f567f8a3819"
+JuliaFormatter = "98e50ef6-434e-11e9-1051-2b60c6c9e899"
```
Comment (charleskawczynski) [inline Project.toml:26]: I think this was accidentally added. Let's remove this.

File: experiments/AtmosLES/bomex_model.jl [hunk 2]
```diff
@@ -511,6 +511,7 @@ function bomex_model(
         problem = problem,
         turbulence = SmagorinskyLilly{FT}(C_smag),
         moisture = moisture,
+        ref_state = NoReferenceState(),
```
Comment (charleskawczynski) [inline experiments/AtmosLES/bomex_model.jl:514]: This will likely result in CI failing. Probably need to revert this. If you need `NoReferenceState` for a specific experiment, maybe we can pass it in as a kwarg and set the `AtmosModel`'s default (`HydrostaticState(DecayingTemperatureProfile{FT}(param_set),)`).

File: experiments/AtmosLES/riemann.jl [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+      state.moisture.ρq_tot = FT(0)
+    end=#
+    ## Test 1
+    if (x < 0)
+        state.ρ = FT(1)
+        state.ρu = SVector{3, FT}(0, 0, 0)
+        p = FT(100000)
+        T = air_temperature_from_ideal_gas_law(param, p, FT(1))
+        state.ρe = internal_energy(param, T)
+        #state.moisture.ρq_tot = FT(0)
```
Comment (charleskawczynski) [inline experiments/AtmosLES/riemann.jl:137]: We should use a thermo state constructor here. We can add a new one, using `ρ` and `p`, if that's what you need and if it makes sense. Same for below.

File: src/Atmos/Model/AtmosModel.jl [hunk 4]
```diff
@@ -898,6 +906,264 @@ function numerical_flux_first_order!(
...
+    e_tot = RoeAverage(ρ⁻, ρ⁺, e⁻, e⁺)
+    h̃ = RoeAverage(ρ⁻, ρ⁺, h⁻, h⁺)
+    qt = RoeAverage(ρ⁻, ρ⁺, qt⁻, qt⁺)
+    ρ = sqrt(ρ⁻ * ρ⁺)
+    e_int⁻ = internal_energy(
+        ρ⁻,
+        ρe⁻,
+        ρu⁻,
+        gravitational_potential(balance_law, state_auxiliary⁻),
+    )
```
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:980]:
```suggestion
    e_int⁻ = internal_energy(ts⁻)
```

File: src/Atmos/Model/AtmosModel.jl [hunk 5]
```diff
@@ -898,6 +906,264 @@ function numerical_flux_first_order!(
...
+    e_int⁺ = internal_energy(
+        ρ⁺,
+        ρe⁺,
+        ρu⁺,
+        gravitational_potential(balance_law, state_auxiliary⁺),
+    )
```
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:986]:
```suggestion
    e_int⁺ = internal_energy(ts⁺)
```

File: src/Atmos/Model/AtmosModel.jl [hunk 6]
```diff
@@ -898,6 +906,264 @@ function numerical_flux_first_order!(
...
+    fluxᵀn::Vars{S},
+    normal_vector::SVector,
+    state_prognostic⁻::Vars{S},
+    state_auxiliary⁻::Vars{A},
+    state_prognostic⁺::Vars{S},
+    state_auxiliary⁺::Vars{A},
+    t,
+    direction,
+) where {S, A}
+
```
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:925]:
```suggestion
    balance_law.moisture isa EquilMoist || error("Must use a EquilMoist model for RoeNumericalFluxMoist")
```

File: src/Atmos/Model/AtmosModel.jl [hunk 7]
```diff
@@ -898,6 +906,264 @@ function numerical_flux_first_order!(
...
+    if (numerical_flux.HH == true)
+        Λ = SDiagonal(
+            max(
+                abs(ũᵀn - c̃_LM),
+                max(
+                    0,
+                    ũᵀn - c̃_LM - (u⁻' * normal_vector - c⁻),
+                    u⁺' * normal_vector - c⁺ - (ũᵀn - c̃_LM),
+                ),
+            ),
```
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:1041]: Can we precompute these components to battle the formatter?

Comment (slifer50) [reply src/Atmos/Model/AtmosModel.jl:1041]: This is written this way to be consistent with the formulation but yes, we could precompute this into separate quantities.

**Comment by slifer50** [comment]
@Zhengyu-Huang Can you have a look at the PR and let me know if there any additional changes necessary, I'm mainly trying to fix a documentation issue, would be good to know if there are any other things I should address.

File: src/Atmos/Model/AtmosModel.jl [hunk 8]
```diff
@@ -1142,4 +1150,251 @@ function numerical_flux_first_order!(
...
+        SVector(0, 0, 0, 0, _e_int_v0, 1),
+    )
+    Δρ = ρ⁺ - ρ⁻
+    Δρu = ρu⁺ - ρu⁻
+    Δρe = ρe⁺ - ρe⁻
+    Δρq_tot = ρq_tot⁺ - ρq_tot⁻
+    Δstate = SVector(Δρ, Δρu[1], Δρu[2], Δρu[3], Δρe, Δρq_tot)
+
+    parent(fluxᵀn) .-= M * Λ * (M \ Δstate) / 2
+    #@info parent(fluxᵀn)
```
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:1397]:
```suggestion
```

File: src/Atmos/Model/linear.jl [hunk 9]
```diff
@@ -388,3 +388,134 @@ function numerical_flux_first_order!(
...
+    ΔpL = pL⁺ - pL⁻
+    Δρuᵀn = (ρu⁺ - ρu⁻)' * normal_vector
+
+    #fluxᵀn.ρ -= ΔpL / 2c̃
+    #fluxᵀn.ρu -= c̃ * Δρuᵀn * normal_vector / 2
+    #fluxᵀn.ρe -= h̃ * ΔpL / 2c̃
+
+    #temp1 = ΔpL / 2c̃
+    #temp2 = c̃ * Δρuᵀn * normal_vector / 2
+    #temp3 = h̃ * ΔpL / 2c̃
```
Comment (charleskawczynski) [inline src/Atmos/Model/linear.jl:480]: Should we delete this?

File: src/Atmos/Model/linear.jl [hunk 10]
```diff
@@ -388,3 +388,134 @@ function numerical_flux_first_order!(
...
+
+
+
+    _T_0::Float64 = T_0(param_set)
+    _e_int_v0 = e_int_v0(param_set)
+    _cv_d = cv_d(param_set)
+    Φ = gravitational_potential(balance_law.atmos, state_auxiliary⁻)
+    # guaranteed to be random
+    ω = Float64(π) / 3
+    δ = Float64(π) / 5
```
Comment (charleskawczynski) [inline src/Atmos/Model/linear.jl:490]: Is there a reason why we're hard-coding `Float64` here?

File: src/Atmos/Model/linear.jl [hunk 11]
```diff
@@ -388,3 +388,134 @@ function numerical_flux_first_order!(
...
+    Δρq_tot =
+        state_prognostic⁺.moisture.ρq_tot - state_prognostic⁻.moisture.ρq_tot
+    Δstate = SVector(Δρ, Δρu[1], Δρu[2], Δρu[3], Δρe, Δρq_tot)
+
+    parent(fluxᵀn) .-= M * Λ * (M \ Δstate) / 2
+    #temp = M * Λ * (M \ Δstate) / 2
+    #temp1 = ΔpL / 2c̃
+    #temp2 = c̃ * Δρuᵀn * normal_vector / 2
+    #temp3 = h̃ * ΔpL / 2c̃
+    #@info temp, temp1, temp2, temp3, state_prognostic⁺.moisture.ρq_tot, state_prognostic⁻.moisture.ρq_tot
```
Comment (charleskawczynski) [inline src/Atmos/Model/linear.jl:520]: Can we delete this if it's not being used?

Comment (slifer50) [reply src/Atmos/Model/linear.jl:520]: Yes, I'll get rid of this, it was just for testing.

File: src/Atmos/Model/linear.jl (hunk 10 above)
Comment (slifer50) [reply src/Atmos/Model/linear.jl:490]: I'll change this to an FT call.

File: src/Atmos/Model/linear.jl (hunk 9 above)
Comment (slifer50) [reply src/Atmos/Model/linear.jl:480]: Yes I'll get rid of this.

**Comment by slifer50** [comment]
@charleskawczynski any idea what might be breaking documentation here? When I run make.jl on while connected to central I don't get any errors.

File: src/Atmos/Model/AtmosModel.jl [hunk 12]
```diff
@@ -1142,4 +1150,250 @@ function numerical_flux_first_order!(
...
+        ρ,
+        qt,
+        balance_law.moisture.maxiter,
+        balance_law.moisture.tolerance,
+    )
+    c̃ = sqrt(roe_average(ρ⁻, ρ⁺, c⁻^2, c⁺^2))
+    _cv_m = cv_m(ts)
+    R_m = gas_constant_air(ts)
+    _cp_m = cp_m(ts)
+    gamma = _cp_m / _cv_m
```
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:1231]:
```suggestion
    (R_m, _cp_m, _cv_m, gamma) = gas_constants(ts)
```

File: src/Atmos/Model/AtmosModel.jl [hunk 13]
```diff
@@ -1142,4 +1150,250 @@ function numerical_flux_first_order!(
...
+
+
+    M = hcat(
+        SVector(1, ũc̃⁺[1], ũc̃⁺[2], ũc̃⁺[3], h̃ - c̃ * ũᵀn, qt),
+        SVector(0, τ1[1], τ1[2], τ1[3], τ1' * ũ, 0),
+        SVector(0, τ2[1], τ2[2], τ2[3], τ2' * ũ, 0),
+        SVector(1, ũ[1], ũ[2], ũ[3], ũ' * ũ / 2 + Φ - _T_0 * _cv_m, 0),
+        SVector(1, ũc̃⁻[1], ũc̃⁻[2], ũc̃⁻[3], h̃ + c̃ * ũᵀn, qt),
+        SVector(0, 0, 0, 0, _e_int_v0, 1),
+    )
```
Comment (Zhengyu-Huang) [inline src/Atmos/Model/AtmosModel.jl:1386]: Have you checked M * Λ * M^{-1} = d(F * n)/dU ? i.e., a unit test.  
This is not efficient (compared with the Dry Roe, OK to live with this now)

File: src/Atmos/Model/AtmosModel.jl [hunk 14]
```diff
@@ -1142,4 +1150,250 @@ function numerical_flux_first_order!(
...
+    _cv_m = cv_m(ts)
+    R_m = gas_constant_air(ts)
+    _cp_m = cp_m(ts)
+    gamma = _cp_m / _cv_m
+    # chosen by fair dice roll
+    # guaranteed to be random
+    ω = FT(π) / 3
+    δ = FT(π) / 5
+    random_unit_vector = SVector(sin(ω) * cos(δ), cos(ω) * cos(δ), sin(δ))
+    # tangent space basis
```
Comment (Zhengyu-Huang) [inline src/Atmos/Model/AtmosModel.jl:1234]: Do any two tangential vectors τ1 τ2 work ? 
Do you need to add assert(  norm(τ1) > 0 ) ? I guess the random_unit_vector is to avoid random_unit_vector // normal_vector ?

File: src/Atmos/Model/linear.jl [hunk 15]
```diff
@@ -388,3 +388,119 @@ function numerical_flux_first_order!(
...
+    ΔpL = pL⁺ - pL⁻
+    Δρuᵀn = (ρu⁺ - ρu⁻)' * normal_vector
+
+    _T_0::FT64 = T_0(param_set)
+    _e_int_v0 = e_int_v0(param_set)
+    _cv_d = cv_d(param_set)
+    Φ = gravitational_potential(balance_law.atmos, state_auxiliary⁻)
+    # guaranteed to be random
+    ω = FT(π) / 3
+    δ = FT64(π) / 5
```
Comment (Zhengyu-Huang) [inline src/Atmos/Model/linear.jl:480]: FT instead of FT64

File: src/Atmos/Model/linear.jl [hunk 16]
```diff
@@ -388,3 +388,119 @@ function numerical_flux_first_order!(
...
+        balance_law,
+        fluxᵀn,
+        normal_vector,
+        state_prognostic⁻,
+        state_auxiliary⁻,
+        state_prognostic⁺,
+        state_auxiliary⁺,
+        t,
+        direction,
+    )
```
Comment (Zhengyu-Huang) [inline src/Atmos/Model/linear.jl:415]: Is this a linearized flux ? why is this flux component nonlinear ? @mwarusz  This is also in the Dry roeflux

File: src/Atmos/Model/linear.jl [hunk 17]
```diff
@@ -388,3 +388,119 @@ function numerical_flux_first_order!(
...
+    ρu⁻ = state_prognostic⁻.ρu
+
+    ref_ρ⁻ = state_auxiliary⁻.ref_state.ρ
+    ref_ρe⁻ = state_auxiliary⁻.ref_state.ρe
+    ref_T⁻ = state_auxiliary⁻.ref_state.T
+    ref_q⁻ = state_auxiliary⁻.ref_state.ρq_tot / ref_ρ⁻
+    ref_qliq⁻ = state_auxiliary⁻.ref_state.ρq_liq / ref_ρ⁻
+    ref_qice⁻ = state_auxiliary⁻.ref_state.ρq_ice / ref_ρ⁻
+    ref_p⁻ = state_auxiliary⁻.ref_state.p
+    #ref_h⁻ = (ref_ρe⁻ + ref_p⁻) / ref_ρ⁻
```
Comment (Zhengyu-Huang) [inline src/Atmos/Model/linear.jl:429]: delete this line #.....

File: src/Atmos/Model/linear.jl [hunk 18]
```diff
@@ -388,3 +388,119 @@ function numerical_flux_first_order!(
...
+    q_pt⁻ = PhasePartition(ref_q⁻, ref_qliq⁻, ref_qice⁻)
```
Comment (Zhengyu-Huang) [inline src/Atmos/Model/linear.jl:430]: is this  q_pt-     ref_q_pt-

File: src/Atmos/Model/linear.jl [hunk 19]
```diff
@@ -388,3 +388,119 @@ function numerical_flux_first_order!(
...
+    ρu⁺ = state_prognostic⁺.ρu
+
+    ref_ρ⁺ = state_auxiliary⁺.ref_state.ρ
+    ref_ρe⁺ = state_auxiliary⁺.ref_state.ρe
+    ref_T⁺ = state_auxiliary⁺.ref_state.T
+    ref_q⁺ = state_auxiliary⁺.ref_state.ρq_tot / ref_ρ⁺
+    ref_qliq⁺ = state_auxiliary⁺.ref_state.ρq_liq / ref_ρ⁺
+    ref_qice⁺ = state_auxiliary⁺.ref_state.ρq_ice / ref_ρ⁺
+    ref_p⁺ = state_auxiliary⁺.ref_state.p
+    #ref_h⁺ = (ref_ρe⁺ + ref_p⁺) / ref_ρ⁺
```
Comment (Zhengyu-Huang) [inline src/Atmos/Model/linear.jl:453]: delete this line # .....

File: src/Atmos/Model/linear.jl [hunk 20]
```diff
@@ -388,3 +388,119 @@ function numerical_flux_first_order!(
...
+    q_pt⁺ = PhasePartition(ref_q⁺, ref_qliq⁺, ref_qice⁺)
```
Comment (Zhengyu-Huang) [inline src/Atmos/Model/linear.jl:454]: q_pt+   ref_q_pt+

File: src/Atmos/Model/linear.jl [hunk 21]
```diff
@@ -388,3 +388,119 @@ function numerical_flux_first_order!(
...
+        atmos.orientation,
+        state_prognostic⁺,
+        state_auxiliary⁺,
+    )
+
+    # not sure if arithmetic averages are a good idea here
+    h̃ = (ref_h⁻ + ref_h⁺) / 2
+    c̃ = (ref_c⁻ + ref_c⁺) / 2
+    qt = (ref_q⁺ + ref_q⁻) / 2
+
```
Comment (Zhengyu-Huang) [inline src/Atmos/Model/linear.jl:470]: add ref

**Review by Zhengyu-Huang — COMMENTED** [review]
Great job! Thank you! But I have some (theoretical) questions.

File: src/Atmos/Model/AtmosModel.jl (hunk 13 above)
Comment (slifer50) [reply src/Atmos/Model/AtmosModel.jl:1386]: Yes the Schur complement has to be calculated for us to make this as efficient as the DryRoe. We can do this at a later time. Would including the flux in the isentropic vortex test work for this purpose?

File: src/Atmos/Model/AtmosModel.jl (hunk 14 above)
Comment (slifer50) [reply src/Atmos/Model/AtmosModel.jl:1234]: Yes pretty much, I don't think norm(τ1) = 0 is a possibility.

File: src/Atmos/Model/AtmosModel.jl [hunk 22]
```diff
@@ -1142,4 +1150,250 @@ function numerical_flux_first_order!(
...
+    e_int = roe_average(ρ⁻, ρ⁺, e_int⁻, e_int⁺)
+    ts = PhaseEquil(
+        param_set,
+        e_int,
+        ρ,
+        qt,
+        balance_law.moisture.maxiter,
+        balance_law.moisture.tolerance,
+    )
+    c̃ = sqrt(roe_average(ρ⁻, ρ⁺, c⁻^2, c⁺^2))
```
[the remark this answers is not in the document]
Comment (slifer50) [reply src/Atmos/Model/AtmosModel.jl:1227]: I'll adjust this to follow your suggestion

File: src/Numerics/DGMethods/NumericalFluxes.jl [hunk 23]
```diff
@@ -341,6 +342,32 @@ Requires a custom implementation for the balance law.
 """
 struct HLLCNumericalFlux <: NumericalFluxFirstOrder end
 
+"""
```
[the remark this answers is not in the document]
Comment (slifer50) [reply src/Numerics/DGMethods/NumericalFluxes.jl:345]: as far as I know the only reference we have for the moist Roe flux is our own documentation thanks to the work by @bischtob and @mwarusz, should we try to list this?

**Comment by slifer50** [comment]
@Zhengyu-Huang I think I've addressed your comments, can you have a look and let me know if there's anything I missed. I've added the flux to the isentropicvortex test. Do you have any other test suggestions?

File: src/Atmos/Model/linear.jl (hunk 16 above)
Comment (slifer50) [reply src/Atmos/Model/linear.jl:415]: I believe this uses the linearized pressure, can you point out which component is nonlinear?

File: src/Numerics/DGMethods/NumericalFluxes.jl [hunk 24]
```diff
@@ -363,6 +390,7 @@ An optional method can also be defined for
                                        Qaux⁻, Q⁺, Qstate_gradient_flux⁺, Qaux⁺, bctype, t)
 
 """
+
```
Comment (kpamnany) [inline src/Numerics/DGMethods/NumericalFluxes.jl:393]:
```suggestion
```

File: src/Atmos/Model/AtmosModel.jl [hunk 25]
```diff
@@ -1141,4 +1149,246 @@ function numerical_flux_first_order!(
...
+                (δ_L_3 + δ_R_3)
+        else
+            qa3 = abs(ũᵀn + c̃_LM)
+        end
+        Λ = SDiagonal(qa2, qa1, qa1, qa1, qa3, qa1)
+    end
+
+
+
+
```
Comment (kpamnany) [inline src/Atmos/Model/AtmosModel.jl:1377]:
```suggestion
```

File: src/Numerics/DGMethods/NumericalFluxes.jl [hunk 26]
```diff
@@ -341,6 +342,32 @@ Requires a custom implementation for the balance law.
...
 struct HLLCNumericalFlux <: NumericalFluxFirstOrder end
 
+"""
+    RoeNumericalFluxMoist <: NumericalFluxFirstOrder
+
+A moist implementation of the numerical flux based on the approximate Riemann solver of Roe
+
+Requires a custom implementation for the balance law.
+"""
+
```
Comment (kpamnany) [inline src/Numerics/DGMethods/NumericalFluxes.jl:352]:
```suggestion
```

**Comment by slifer50** [comment]
@Zhengyu-Huang @charleskawczynski @mwarusz I think this is about ready to be merged, are there any concerns that still need to be addressed before we can r+ this?

**Review by Zhengyu-Huang — APPROVED** [review]
That looks great! Thank you! Like the new test!

**Comment by slifer50** [comment]
Yes this should apply the central flux the linearized balance law

**Comment by slifer50** [comment]
I've just rebased this, @charleskawczynski @Zhengyu-Huang let's r+ if there are no objections

File: test/Numerics/DGMethods/Euler/isentropicvortex.jl [hunk 27]
```diff
@@ -222,15 +243,19 @@ function test_run(
...
+    end
     model = AtmosModel{FT}(
         AtmosLESConfigType,
         param_set;
         problem = problem,
         orientation = NoOrientation(),
         ref_state = NoReferenceState(),
         turbulence = ConstantDynamicViscosity(FT(0)),
-        moisture = DryModel(),
+        moisture = moisture,#DryModel(),
```
Comment (charleskawczynski) [inline test/Numerics/DGMethods/Euler/isentropicvortex.jl:258]:
```suggestion
        moisture = moisture,
```

File: src/Atmos/Model/linear.jl [hunk 28]
```diff
@@ -388,3 +388,124 @@ function numerical_flux_first_order!(
...
+    ref_h̃ = (ref_h⁻ + ref_h⁺) / 2
+    ref_c̃ = (ref_c⁻ + ref_c⁺) / 2
+    ref_qt = (ref_q⁺ + ref_q⁻) / 2
+
+    ΔpL = pL⁺ - pL⁻
+    Δρuᵀn = (ρu⁺ - ρu⁻)' * normal_vector
+
+    _T_0::FT = T_0(param_set)
+    _e_int_v0 = e_int_v0(param_set)
+    _cv_d = cv_d(param_set)
```
Comment (charleskawczynski) [inline src/Atmos/Model/linear.jl:474]:
```suggestion
    _e_int_v0::FT = e_int_v0(param_set)
    _cv_d::FT = cv_d(param_set)
```

File: src/Atmos/Model/AtmosModel.jl [hunk 29]
```diff
@@ -1141,4 +1149,243 @@ function numerical_flux_first_order!(
...
+        t,
+        direction,
+    )
+
+    FT = eltype(fluxᵀn)
+    param_set = balance_law.param_set
+    _cv_d::FT = cv_d(param_set)
+    _T_0::FT = T_0(param_set)
+    γ::FT = cp_d(param_set) / cv_d(param_set)
+    _e_int_v0 = e_int_v0(param_set)
```
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:1184]:
```suggestion
    _e_int_v0::FT = e_int_v0(param_set)
```

File: src/Atmos/Model/AtmosModel.jl [hunk 30]
```diff
@@ -1141,4 +1149,243 @@ function numerical_flux_first_order!(
...
+    e_int = roe_average(ρ⁻, ρ⁺, e_int⁻, e_int⁺)
+    ts = PhaseEquil(
+        param_set,
+        e_int,
+        ρ,
+        qt,
+        balance_law.moisture.maxiter,
+        balance_law.moisture.tolerance,
+    )
+    c̃ = sqrt((γ - 1) * (h̃ - (ũ[1]^2 + ũ[2]^2 + ũ[3]^2) / 2))#sqrt(roe_average(ρ⁻, ρ⁺, c⁻^2, c⁺^2))
```
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:1226]:
```suggestion
    c̃ = sqrt((γ - 1) * (h̃ - (ũ[1]^2 + ũ[2]^2 + ũ[3]^2) / 2))
```

File: src/Atmos/Model/AtmosModel.jl [hunk 31]
```diff
@@ -1141,4 +1149,243 @@ function numerical_flux_first_order!(
...
+    ũᵀn⁻ = u⁻' * normal_vector
+    ũᵀn⁺ = u⁺' * normal_vector
+    ũᵀn = ũ' * normal_vector
+    ũc̃⁻ = ũ + c̃ * normal_vector
+    ũc̃⁺ = ũ - c̃ * normal_vector
+    e_kin_pot = h̃ - _e_int_v0 * qt - _cp_m * c̃^2 / R_m
+    if (numerical_flux.LM == true)
+        Mach⁺ = sqrt(u⁺' * u⁺) / c⁺
+        Mach⁻ = sqrt(u⁻' * u⁻) / c⁻
+        Mach = (Mach⁺ + Mach⁻) / 2 #roe_average(ρ⁻, ρ⁺, Mach⁻, Mach⁺)
```
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:1245]:
```suggestion
        Mach = (Mach⁺ + Mach⁻) / 2
```

**Comment by slifer50** [comment]
I've rebased this again.

**Review by charleskawczynski — COMMENTED** [review]
Can we add tests for when `LM, HH, LV, LVPP` are true (I'm assuming one at a time?)

**Comment by slifer50** [comment]
> Can we add tests for when `LM, HH, LV, LVPP` are true (I'm assuming one at a time?)

I can try to include this in the isentropic vortex test, I'll get to work on it.

**Comment by slifer50** [comment]
@charleskawczynski I've added in the LM, etc tests to the isentropic vortex case.

**Comment by slifer50** [comment]
Is this in a mergeable state ? @charleskawczynski

File: test/Numerics/DGMethods/Euler/isentropicvortex.jl [hunk 32]
```diff
@@ -379,6 +498,9 @@ function isentropicvortex_initialcondition!(
     state.ρu = ρ * u
     e_kin = u' * u / 2
     state.ρe = ρ * total_energy(e_kin, e_pot, ts)
+    if bl.moisture isa EquilMoist
```
Comment (charleskawczynski) [inline test/Numerics/DGMethods/Euler/isentropicvortex.jl:501]:
```suggestion
    if !(bl.moisture isa DryModel)
```

File: test/Numerics/DGMethods/Euler/isentropicvortex.jl [hunk 33]
```diff
@@ -222,22 +337,26 @@ function test_run(
...
-        moisture = DryModel(),
+        moisture = moisture,
         source = (),
     )
 
     dg = DGModel(
         model,
         grid,
-        NumericalFlux(),
+        NumericalFlux,#(),
```
Comment (charleskawczynski) [inline test/Numerics/DGMethods/Euler/isentropicvortex.jl:359]:
```suggestion
        NumericalFlux,
```

**Review by charleskawczynski — APPROVED** [review]
LGTM. Let's squash and merge?

**Comment by slifer50** [comment]
rebase?

**Comment by slifer50** [comment]
Squashed and good to go

File: src/Atmos/Model/AtmosModel.jl [hunk 34]
```diff
@@ -1141,4 +1149,243 @@ function numerical_flux_first_order!(
...
+    Λ = SDiagonal(
+        abs(ũᵀn - c̃_LM),
+        abs(ũᵀn),
+        abs(ũᵀn),
+        abs(ũᵀn),
+        abs(ũᵀn + c̃_LM),
+        abs(ũᵀn),
+    )
+    #Harten Hyman
+    if (numerical_flux.HH == true)
```
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:1260]: @slifer50 just to note, it might be an improvement to use dispatch for these, rather than `Bool`s. But the method interface might not be the same and some parts are needed for multiple numerical fluxes, so I couldn't see a clear win at first glance.

Comment (slifer50) [reply src/Atmos/Model/AtmosModel.jl:1260]: Yeah, I'm not a big fan of the current way this is being handled, but I don't really know a simple way to do this without building these as entirely separate fluxes.

Comment (charleskawczynski) [reply src/Atmos/Model/AtmosModel.jl:1260]: We can try to do that, if it's worthwhile, in a follow-up PR

Comment (slifer50) [reply src/Atmos/Model/AtmosModel.jl:1260]: I agree with this.
